Diagnostics

Steve-Mcl opened the pull request:
fixes #3430

## Types of changes

- [x] New feature (non-breaking change which adds functionality)


## Proposed changes

REF: #3430

### Notable Changes

#### Backend
* adds admin endpoint `/diagnostics`
* adds new settings `diagnostics: { enabled: bool, ui: bool }`


#### UI
* adds action `core:show-system-info`
* adds download and copy buttons to JSON editor
* adds `readonly` option to JSON editor

### TODO
* i18n - necessary?
* ~~tests~~

### Demo
![chrome_8cVfY6mYKB](https://user-images.githubusercontent.com/44235289/160673602-c74d908d-4826-49a0-bfa6-1d6c2624c32d.gif)



## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [x] I have added suitable unit tests to cover the new/changed functionality

knolleary commented on packages/node_modules/node-red/lib/red.js:
```diff
@@ -229,5 +229,12 @@ module.exports = {
+     * @see @node-red/editor-api_auth
+     * @memberof node-red
+     */
+    get diagnostics() { debugger; return api.diagnostics }
```
```suggestion
    get diagnostics() { return api.diagnostics }
```

knolleary commented on packages/node_modules/@node-red/editor-client/src/js/ui/editors/json.js:
```diff
@@ -498,7 +507,25 @@
+                                .on("click", function (evt) {
+                                    evt.preventDefault();
+                                    if (button.click !== undefined) {
+                                        button.click(evt);
```
```suggestion
                                        button.click.call(element, evt);
```
Need to call the function with `this` set to the actual button element - so that `$(this)` evaluates to the right thing. Otherwise you get an error in the console when clicking the Copy button.

knolleary commented on packages/node_modules/@node-red/editor-client/src/js/ui/diagnostics.js:
```diff
@@ -0,0 +1,61 @@
+                    json = "{\n\n}";
+                }
+                RED.editor.editJSON({
+                    title: "System Info", //RED._('sidebar.project.editDependencies'),
```
Once the other proposed changes are added - I think this is the final piece of text to pull out into a catalog.

Steve-Mcl replied:
Done in 367f9b6

knolleary reviewed (commented):
Some i18n updates - using existing messages.

One other tweak, but as it touches lines of code that you haven't touched, I cannot propose changes on them.

But in `editors/json.js` we should wrap the `node-input-json-reformat` button in a `button-group` span - this will ensure that as more buttons are added, they are presented properly:

```
                '<span class="button-group">'+
                    '<button id="node-input-json-reformat" class="red-ui-button red-ui-button-small"><span data-i18n="jsonEditor.format"></span></button>'+
                '</span>'+
```

Current:

<img width="275" alt="image" src="https://user-images.githubusercontent.com/51083/165296353-006a5c8e-8820-488d-a571-db3854d4db64.png">


With the button-group:

<img width="284" alt="image" src="https://user-images.githubusercontent.com/51083/165296294-b952968f-94fd-4a46-9560-90cf85dae645.png">

Steve-Mcl commented:
@knolleary 

* Added missing i18n
* renamed internal var `jsonButtons` in the JSON Editor options to `toolbarButtons` (makes it more accessible) 
* grouped json editor toolbar buttons as requested.

Ready for final review.
